frontend/catalyst/utils/toml.py

```diff
@@ -369,7 +424,107 @@ def get_device_capabilities(
         to_matrix_ops=matrix_decomp_props,
         native_obs=observable_props,
         measurement_processes=measurements_props,
-        mid_circuit_measurement_flag=check_compilation_flag(config, "mid_circuit_measurement"),
-        runtime_code_generation_flag=check_compilation_flag(config, "runtime_code_generation"),
-        dynamic_qubit_management_flag=check_compilation_flag(config, "dynamic_qubit_management"),
+        qjit_compatible_flag=get_compilation_flag(config, "qjit_compatible"),
+        mid_circuit_measurement_flag=get_compilation_flag(config, "mid_circuit_measurement"),
+        runtime_code_generation_flag=get_compilation_flag(config, "runtime_code_generation"),
+        dynamic_qubit_management_flag=get_compilation_flag(config, "dynamic_qubit_management"),
+        options=get_options(config),
     )
+
+
+def check_no_overlap(*args, device_name):
+    """Check items in *args are mutually exclusive.
+
+    Args:
+        *args (List[Str]): List of strings.
+        device_name (str): Device name for error reporting.
+
+    Raises:
+        CompileError
+    """
+    set_of_sets = [set(arg) for arg in args]
+    union = set.union(*set_of_sets)
+    len_of_sets = [len(arg) for arg in args]
+    if sum(len_of_sets) == len(union):
+        return
+
+    overlaps = set()
+    for s in set_of_sets:
+        overlaps.update(s - union)
+        union = union - s
+
+    msg = f"Device '{device_name}' has overlapping gates: {overlaps}"
+    raise CompileError(msg)
+
+
+def filter_out_adjoint(operations):
+    """Remove Adjoint from operations.
+
+    Args:
+        operations (List[Str]): List of strings with names of supported operations
+
+    Returns:
+        List: A list of strings with names of supported operations with Adjoint and C gates
+        removed.
+    """
+    adjoint = re.compile(r"^Adjoint\(.*\)$")
+
+    def is_not_adj(op):
+        return not re.match(adjoint, op)
+
+    operations_no_adj = filter(is_not_adj, operations)
+    return set(operations_no_adj)
+
+
+def validate_device_capabilities(
```

Comment:
This module is just here to abstract away interacting with a TOML file right (and abstracting away from schema differences)? Validating device capabilities once the toml file has been processed then seems unrelated to interacting with the toml format. In that case I would prefer having this in the `catalyst/device` submodule, along with some of the other functions that don't interact with the toml file (I'm not 100% where the boundary should be the, probably anything that requires the device instance).

Reply:
Done.
I personally think that it is not possible to find a perfect locations for code in Python+OOP style, so I am trying to focus on observable things only, like cyclic dependencies. The problem is in the fact that in this style we define types and methods simultaneously when we define classes. For my projects I (1) don't use "normal" classes (2) create `types.py` for types and dataclases (3) group functions in -py files according to their semantics.

Reply:
What about get_device_toml_config and get_device_capabilities, their code seems to be operating primarily on the device, and the calls into toml functions like read_toml_file and load_device_capabilities?

Reply:
Done

Reply:
Where?